File: zad1.py
import random

def Z_1():
    N = 7
    print("Задание 1_____________________________________________________________________________________________\n\n")
    matrix = [[random.randint(-10, 10) for _ in range(N)] for _ in range(N)]
    newMatrix = [[0 for _ in range(N)] for _ in range(2)]

    for i in range(N):
        newMatrix[0][i] = matrix[i][i]
        newMatrix[1][i] = matrix[N - 1 - i][i]

    print("Исходная матрица:")
    for i in range(N):
        print("\t| ", end="")
        for j in range(N):
            print(matrix[i][j], end=" ")
        print("\t|")

    print("\n\nНовая матрица:")
    for i in range(2):
        print("\t| ", end="")
        for j in range(N):
            print(newMatrix[i][j], end=" ")
        print("\t|")

    print("\n\n")
    return 0


def Z_2():
    print("Задание 2_____________________________________________________________________________________________\n\n")
    random.seed()
    M = 5
    N = 5
    matrix = [[0 for _ in range(N)] for _ in range(M)]

    print("Матрица:")
    for i in range(M):
        print("\t| ", end="")
        for j in range(N):
            matrix[i][j] = random.randint(-100, 100)
            print(matrix[i][j], end=" ")
        print("\t|")

def find_max_and_second_min(matrix):
    max_element = float('-inf')
    second_min_element = float('inf')

    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            if matrix[i][j] > max_element:
                second_min_element = max_element
                max_element = matrix[i][j]
            elif matrix[i][j] > second_min_element and matrix[i][j] < max_element:
                second_min_element = matrix[i][j]

    return max_element, second_min_element


def replace_diagonal_elements(matrix, max_element, second_min_element):
    for i in range(min(len(matrix), len(matrix[0]))):
        matrix[i][i] = max_element - second_min_element
        matrix[i][len(matrix[0]) - i - 1] = max_element - second_min_element

    return matrix


def print_matrix(matrix):
    print("\nПреобразованная матрица:")
    for i in range(len(matrix)):
        print("\t| ", end="")
        for j in range(len(matrix[0])):
            print(matrix[i][j], end=" ")
        print("\t|")


def is_safe(grid, row, col, num):
    for i in range(N):
        if grid[row][i] == num or grid[i][col] == num:
            return False

    start_row = row - row % 3
    start_col = col - col % 3
    for i in range(start_row, start_row + 3):
        for j in range(start_col, start_col + 3):
            if grid[i][j] == num:
                return False

    return True


def is_valid_sudoku(grid):
    for i in range(N):
        for j in range(N):
            if grid[i][j] != 0:
                num = grid[i][j]
                grid[i][j] = 0
                if not is_safe(grid, i, j, num):
                    return False
                grid[i][j] = num

    return True


if __name__ == "__main__":
    Z_1()
    Z_2()
    matrix = [
        [5, 3, 0, 0, 7, 0, 0, 0, 0],
        [6, 0, 0, 1, 9, 5, 0, 0, 0],
        [0, 9, 8, 0, 0, 0, 0, 6, 0],
        [8, 0, 0, 0, 6, 0, 0, 0, 3],
        [4, 0, 0, 8, 0, 3, 0, 0, 1],
        [7, 0, 0, 0, 2, 0, 0, 0, 6],
        [0, 6, 0, 0, 0, 0, 2, 8, 0],
        [0, 0, 0, 4, 1, 9, 0, 0, 5],
        [0, 0, 0, 0, 8, 0, 0, 7, 9]
    ]

    max_element, second_min_element = find_max_and_second_min(matrix)
    transformed_matrix = replace_diagonal_elements(matrix, max_element, second_min_element)
    print_matrix(transformed_matrix)

    if is_valid_sudoku(matrix):
        print("True")
    else:
        print("False")

File: zad1.cpp
#include <iostream>
#include <vector>
#include <random>
#include <cstdlib>
#include <ctime>
#include <unordered_set>
#include <cstdlib>
using namespace std;

int Z_1() {
    const int N = 7; // Размерность матрицы (задаем N = 7)
    
    // Создаем и инициализируем исходную матрицу размерности N случайными числами
    vector<vector<int>> matrix(N, vector<int>(N));
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> dist(-10, 10); // диапазон случайных чисел от -10 до 10

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            matrix[i][j] = dist(gen);
        }
    }

    // Создаем новую матрицу размерности 2xN для хранения копий элементов диагоналей
    vector<vector<int>> newMatrix(2, vector<int>(N));

    // Заполняем новую матрицу копиями элементов диагоналей исходной матрицы
    for (int i = 0; i < N; ++i) {
        newMatrix[0][i] = matrix[i][i]; // Копируем элементы главной диагонали
        newMatrix[1][i] = matrix[N - 1 - i][i]; // Копируем элементы побочной диагонали
    }

    // Выводим исходную матрицу
    
    for (int i = 0; i < N; ++i) {
        cout << "\t| ";
        for (int j = 0; j < N; ++j) {
            cout << matrix[i][j] << " ";
        }
        cout << "\t|" << endl;
    }
    cout << endl;
    
    // Выводим новую матрицу
   
    for (int i = 0; i < 2; ++i) { // выводим две строки новой матрицы
        cout << "\t| ";
        for (int j = 0; j < N; ++j) {
            cout << newMatrix[i][j] << " ";
        }
        cout << "\t|" << endl;
    }
    cout << endl << endl;
    return 0;
}


int Z_2() {
    setlocale(LC_ALL, "ru");
    
    srand(time(NULL)); // Инициализация генератора псевдослучайных чисел

    int M = 5; // Количество строк матрицы
    int N = 5; // Количество столбцов матрицы

    vector<vector<int>> matrix(M, vector<int>(N)); // Объявление матрицы размерности M x N

    // Инициализация матрицы случайными целыми числами

    for (int i = 0; i < M; ++i) {
        cout << "\t| ";
        for (int j = 0; j < N; ++j) {
            matrix[i][j] = rand() % 201 - 100; // Генерация случайного числа от -100 до 100
            cout << matrix[i][j] << " ";
        }
        cout << "\t|" << endl;
        
    }

    // Нахождение наибольшего и второго наименьшего элементов в матрице
    int maxElement = INT_MIN;
    int secondMinElement = INT_MAX;

    for (int i = 0; i < M; ++i) {
        for (int j = 0; j < N; ++j) {
            if (matrix[i][j] > maxElement) {
                secondMinElement = maxElement;
                maxElement = matrix[i][j];
            }
            else if (matrix[i][j] > secondMinElement && matrix[i][j] < maxElement) {
                secondMinElement = matrix[i][j];
            }
        }
    }

    // Замена диагональных элементов разностью наибольшего и второго наименьшего элементов
    for (int i = 0; i < min(M, N); ++i) {
        matrix[i][i] = maxElement - secondMinElement;
        matrix[i][N - i - 1] = maxElement - secondMinElement;
    }

    // Печать преобразованной матрицы
    
    for (int i = 0; i < M; ++i) {
        cout << "\t| ";
        for (int j = 0; j < N; ++j) {
            cout << matrix[i][j] << " ";
        }
        cout << "\t| ";
        cout << endl;
    }

    return 0;
}

const int N = 9;

bool isSafe(vector<vector<int>>& grid, int row, int col, int num) {
    for (int i = 0; i < N; i++) {
        if (grid[row][i] == num || grid[i][col] == num)
            return false;
    }

    int startRow = row - row % 3;
    int startCol = col - col % 3;
    for (int i = startRow; i < startRow + 3; i++) {
        for (int j = startCol; j < startCol + 3; j++) {
            if (grid[i][j] == num)
                return false;
        }
    }
    return true;
}

bool isValidSudoku(vector<vector<int>>& grid) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (grid[i][j] != 0) {
                int num = grid[i][j];
                grid[i][j] = 0; // временно очищаем клетку для проверки
                if (!isSafe(grid, i, j, num)) {
                    return false;
                }
                grid[i][j] = num; // восстанавливаем значение клетки
            }
        }
    }
    return true;
}




int main() {
    setlocale(LC_ALL, "ru");
    Z_1(); 
    Z_2(); 
    cout << endl;
    vector<vector<int>> grid = {
        {5, 3, 0, 0, 7, 0, 0, 0, 0},
        {6, 0, 0, 1, 9, 5, 0, 0, 0},
        {0, 9, 8, 0, 0, 0, 0, 6, 0},
        {8, 0, 0, 0, 6, 0, 0, 0, 3},
        {4, 0, 0, 8, 0, 3, 0, 0, 1},
        {7, 0, 0, 0, 2, 0, 0, 0, 6},
        {0, 6, 0, 0, 0, 0, 2, 8, 0},
        {0, 0, 0, 4, 1, 9, 0, 0, 5},
        {0, 0, 0, 0, 8, 0, 0, 7, 9}
    };

    if (isValidSudoku(grid)) {
        cout << "True" << endl;
    }
    else {
        cout << "False" << endl;
    }

    return 0;

   
}

